src/test/java/com/diffmin/AppTest.java

```diff
@@ -73,22 +72,30 @@
          * @return instance of {@link TestResources}
          */
         public static TestResources fromTestDirectory(File testDir) {
-            Path path = testDir.toPath();
             String parent = testDir.getName();
-            Stream<Path> paths = Arrays.stream(path.toFile().listFiles()).map(File::toPath);
-            List<Path> pathList = paths.collect(Collectors.toList());
-            Map<String, Path> testPathMap = new HashMap<>();
-            for (Path testPath : pathList) {
-                if (testPath.getFileName().toString().startsWith(AppTest.PREV_PREFIX)) {
-                    testPathMap.put(AppTest.PREV_PREFIX, testPath);
-                } else if (testPath.getFileName().toString().startsWith(AppTest.NEW_PREFIX)) {
-                    testPathMap.put(AppTest.NEW_PREFIX, testPath);
-                }
-            }
-            return new TestResources(
-                    testPathMap.get(AppTest.PREV_PREFIX),
-                    testPathMap.get(AppTest.NEW_PREFIX),
-                    parent);
+            Path prevPath = getFilepathByPrefix(testDir, PREV_PREFIX);
+            Path newPath = getFilepathByPrefix(testDir, NEW_PREFIX);
+            return new TestResources(prevPath, newPath, parent);
+        }
+
+        /**
+         * Returns test resource with the matching prefix.
+         *
+         * @param dir Directory where the test resources are located
+         * @param prefix Prefix of the test resource
+         * @return {@link Path} to the test resource
+         */
```

Comment:
Two notes here:

First, I personally wouldn't add a javadoc to a private method like this. The name of the method is probably sufficient for most to figure out what it does, and if someone is uncertain the method is _right there in the same file_, so they can just have a look at it (and it's very small).

Second, I wouldn't add types in the javadoc (i.e. `{@link Path}`. We know the type of the return value, it's in the method header, and it's automatically included in the rendered documentation. So, it duplicates information. Perhaps you later decide to change the return type to `File`, and then forget to edit the `@return` tag. The tags should describe the meaning or purpose of the values, not their types.

Reply:
> The name of the method is probably sufficient for most to figure out what it does

I agree with that but I write docstrings just to clarify further what does the function exactly does. I feel that it is always easier to read a piece of text than the code. Probably, this logic applies better when we are writing APIs for consumers but I just extend this here as well :)

> Second, I wouldn't add types in the javadoc (i.e. {@link Path}

Yes, I will take a note of this. I could have changed it if you did not merge the PR but anyway, I will do it later if I ever come to this function again.

Reply:
> I agree with that but I write docstrings just to clarify further what does the function exactly does. I feel that it is always easier to read a piece of text than the code.

When possible, let code speak for itself. If you have to explain the code, consider using more descriptive names. A common saying is that "comments are deodorant for code smells" (this isn't necessarily true as a blanket statement, but you get the idea).

There are a couple of very big differences between documenting public and private methods. A public method is part of the class' API, and so we're unlikely to change it often (because that would break dependents of that class). A private method on the other hand is not part of the API, and so we can change it however we wish without affecting anything outside the class itself. Therefore, private methods are more prone to being changed, and elaborate method comments thus more prone to going out of date.

The second big difference is that anyone using the private method must be editing the class itself, and thus always has access to the source code. Thus, in combination with good naming, there's rarely a reason to comment private methods, and if there is, consider just adding a minimal explanation:

```java
        /** Returns a test resource with the matching prefix in the given directory. */
```

There are of course exceptions, sometimes private methods are complicated by necessity and deserve the same level of documentation as public ones. But in general, I find that private methods can go entirely without documentation, or with a minimal explanation, I rarely write a full-blown Javadoc for them. This is of course just my opinion, there's no one truth here.

> Yes, I will take a note of this. I could have changed it if you did not merge the PR but anyway.

You could have, but it's not worth your time. Even if it were to go out of date for this test method, it doesn't really matter. Again, we can't be perfectionists when it comes to code, it's simply not time efficient. This is overall a very tidy PR, and there's nothing objectively wrong with it, just some subjective things.

> I will do it later if I ever come to this function again.

Exactly. I bet this won't happen, though :)

Reply:
> 1.  A public method is part of the class' API, and so we're unlikely to change
> 2. The second big difference is that anyone using the private method must be editing the class itself, and thus always has access to the source code

I am amazed how you always have such concrete logic about the very minute details of Java programming language or any language in general. Thanks for always being as detailed as possible and not terse at all. It really helps a lot! :)

> I bet this won't happen, though :)

I had a similar emotion while writing the comment. 😅